---
title: "Samples with min 10x coverage with P. sojae as an outgroup"
author: "Shankar K Shakya"
date: "December 10, 2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(pals)
```

## Preparing vcf file

```{r, eval=FALSE, echo=TRUE}

rm(list = ls())
library(vcfR, quietly = T)
library(poppr, quietly = T)
library(ggplot2, quietly = T)
library(reshape2, quietly = T)
library(mypackage)

good_variants <- read.vcfR("Good_895variants.gz", verbose = FALSE)

raw_vcf <- read.vcfR("Pcinna242.rmdup.gvcf2vcf.vcf.gz", verbose = FALSE)
raw_vcf <- subset_vcf2vcf(raw_vcf, good_variants)

Psojae <- read.vcfR("Psojae.vcf")
Psojae <- subset_vcf2vcf(Psojae, good_variants)

vcf <- raw_vcf[ , -(grep("Rep2.fq|Rep3.fq|254-3579-WAus.fq", colnames(raw_vcf@gt)))]

dp <- extract.gt(vcf, element = "DP", as.numeric=TRUE)
myMiss <- apply(dp, MARGIN = 2, function(x){ sum( is.na(x))} ) # number of NA per sample
myMiss <- myMiss / nrow(dp)
vcf@gt <- vcf@gt[, c(TRUE, myMiss < 0.1)]

dp <- extract.gt(vcf, element = "DP", as.numeric=TRUE)

check <- colMeans(dp, na.rm = T) >= 10
#ncol(vcf@gt[, c(TRUE, check)])
vcf@gt <- vcf@gt[, c(TRUE, check)]

vcf


vcf@gt <- cbind(vcf@gt, Psojae@gt[, 2])


#write.vcf(vcf, "Min10x_cov_205isolates_895Variants.gz")



```



## Reading prepared vcf file.

Total isolates 203 with 1027 variants (1010 SNPS, 17 indels) with P.sojae as an outgroup

```{r, fig.align="center", fig.height=12, fig.width=12}

library(vcfR, quietly = T)
library(poppr, quietly = T)
library(ggplot2, quietly = T)
library(reshape2, quietly = T)

vcf <- read.vcfR("Min10x_cov_205isolates_895Variants.gz")

#vcf <- extract.indels(vcf)

dp <- extract.gt(vcf, element = "DP", as.numeric=TRUE)
par(mar=c(12,4,4,2))
#boxplot(dp, col=2:8, las=3)
boxplot(dp, col=2:8, xaxt="n")
title(ylab = "Depth (DP)")


id <- unlist(strsplit(colnames(vcf@gt)[-1], split = ".fq"))
pcinna_pop <- read.csv("New Microsoft Excel Worksheet.csv", header = TRUE)
pcinna_pop <- pcinna_pop[pcinna_pop$Isolate %in% id, ]
pcinna_pop <- pcinna_pop[match(id, pcinna_pop$Isolate), ]

pop_cinna <- pcinna_pop$Country

pop_cinna_sojae <- as.factor(c(as.character(pop_cinna), "Psojae"))

vcf_cinna_sojae <- vcf
colnames(vcf_cinna_sojae@gt)[length(colnames(vcf_cinna_sojae@gt))] <- "Psojae"

vcf_cinna <- vcf[ , -ncol(vcf_cinna_sojae@gt)]


```

## Haplotype diversity and number of unique haplotypes

```{r, fig.align="center", fig.height=12, fig.width=12}

#vcf_cinna
#pop_cinna

gt_cinna <- extract.gt(vcf_cinna, element = "GT", return.alleles = TRUE)
gt_cinna <- alleles2consensus(gt_cinna)
gt_cinna <- t(gt_cinna)

library(ape)
gt_cinna_dnabin <- as.DNAbin(gt_cinna)
rownames(gt_cinna_dnabin) <- paste("Pop", pop_cinna, sep = "_")

library(pegas)
h <- haplotype(gt_cinna_dnabin)
h <- sort(h, what = "label")

h.freq <- haploFreq(gt_cinna_dnabin, haplo = h)

rownames(h.freq) <- paste("Haplotype", 1:nrow(h.freq), sep = "_")

h.freq.df <- as.data.frame(h.freq)
h.freq.mat <- h.freq


# barplot(colSums(h.freq > 0) / colSums(h.freq)) ##Haplotypic diversity
# barplot(colSums(h.freq.df > 0) /17) ##Haplotypic diversity corrected for sample size 


hap.div <- as.data.frame(colSums(h.freq.df > 0) /17)
colnames(hap.div) <- "Haplotype_diversity"


#tiff("./MANUSCRIPT_FIGS/Haplotype_diversity_proabablity.tiff", width = 7, height = 7, units = "in", res = 600)

ggplot(data = hap.div,  aes(x = rownames(hap.div), y =  Haplotype_diversity, fill = rownames(hap.div))) + 
  geom_bar(stat = "identity") +
  theme(legend.position="none") +
  theme(legend.text = element_text(size = 10)) +
  theme(axis.text.x = element_text(size = 10, face = "bold", angle = 60, hjust = 1)) +
  theme(axis.text.y = element_text(size = 10, face = "bold")) +
  labs(x = "Population", y = "Haplotype diversity") +
  theme(axis.title = element_text(size = 15, face = "bold")) +
  theme(axis.text.y = element_text(size = 10, face = "bold")) 

#dev.off()

```

## Haplonet plots

```{r}

nt <- haploNet(h)
fq <- attr(nt, "freq")

#tiff("./MANUSCRIPT_FIGS/Haplotype_diversity.tiff", width = 7, height = 7, units = "in", res = 600)

plot(nt, size = 4,labels =F, lwd = 0.5, asp =1.5, fast = T, 
     show.mutation = F, threshold = 0, pie = h.freq.mat, bg = alphabet(n = ncol(h.freq)))

legend(-31, 13, colnames(h.freq), fill=alphabet(n = ncol(h.freq)), 
       cex=1, horiz = FALSE, ncol = 1, box.col = "white", bg = "transparent", bty = "n")

#dev.off()


#tiff("./MANUSCRIPT_FIGS/Haplotype_diversity2.tiff", width = 7, height = 7, units = "in", res = 600)

plot(nt, size = attr(nt, "freq"),labels =F, lwd = 0.5, asp =1.5, fast = T, 
     show.mutation = F, threshold = 0, pie = h.freq.mat, bg = alphabet(n = ncol(h.freq)))

legend(-145, 55, colnames(h.freq), fill=alphabet(n = ncol(h.freq)), 
       cex=1, horiz = FALSE, ncol = 1, box.col = "white", bg = "transparent", bty = "n")

#dev.off()

```

## Number of unique haplotypes

```{r}

head(h.freq.df) 

uniq_haplotypes <- h.freq.df[rowSums(h.freq.df > 0) == 1 , ]

colSums(uniq_haplotypes > 0) 

## 8 Vietnam
## 2 Taiwan
## 1 South Africa


```



## Multilocus genotype and shared MLGs

```{r, echo=TRUE, fig.align="center", fig.width=12, fig.height=12}

vcf.genind <- vcfR2genind(vcf_cinna)
pop(vcf.genind) <- pop_cinna

tab <- mlg.table(vcf.genind)
mlg.crosspop(vcf.genind)


```

## Neighbor joining tree

```{r, fig.align="center", fig.height=12, fig.width=12}
library(ggtree)
library(ggrepel)
library(poppr)
library(dplyr)
library(adegenet)
library(ape)

vcf.gl <- vcfR2genlight(vcf_cinna)

tree_nj <- aboot(vcf.gl, tree = "nj", distance = bitwise.dist, sample = 1000, showtree = F, cutoff = 50)

library(phangorn, quietly = TRUE)
tree_nj <- midpoint(tree_nj)


countryInfo <- split(tree_nj$tip.label, pop_cinna)
tree2 <- groupOTU(tree_nj, countryInfo)

tree2$tip.label <- paste(pcinna_pop$Country_code, pcinna_pop$MT, sep = "_")
# 
ggtree(tree2, aes(color=group, label = node), layout="circular", ladderize = TRUE) +
  geom_tiplab(size=3, aes(angle=angle))

#ggtree(tree2, aes(color=group, label = node), ladderize = TRUE) + geom_tiplab(size=3) 

# ggtree(tree2, aes(color=group)) +
#   geom_tiplab(size=3) +
#   geom_text2(aes(subset=!isTip, label=node), hjust=-0.3) + geom_treescale()



## With PSojae

vcf.gl_sojae <- vcfR2genlight(vcf_cinna_sojae)

tree_nj_sojae <- aboot(vcf.gl_sojae, tree = "nj", distance = bitwise.dist, sample = 1000, showtree = F, cutoff = 50)

library(phangorn, quietly = TRUE)

tree_nj_sojae <- ape::root(tree_nj_sojae, outgroup = "Psojae")


countryInfo <- split(tree_nj_sojae$tip.label, pop_cinna_sojae)
tree2_sojae <- groupOTU(tree_nj_sojae, countryInfo)


tree2_sojae$tip.label <- paste(pcinna_pop$Country_code, pcinna_pop$MT, sep = "_")
tree2_sojae$tip.label[205] <- "Psojae"

ggtree(tree2_sojae, aes(color=group, label = node), layout="circular", ladderize = TRUE) +
  geom_tiplab(size=3, aes(angle=angle))

# ggtree(tree2_sojae, aes(color=group, label = node), ladderize = TRUE) +
#   geom_tiplab(size=3)

```


## convert to genlight and Minimum spanning network

```{r, eval=F}

vcf.gl <- vcfR2genlight(vcf_cinna)
#pop(vcf.gl) <- pop[-length(pop)]
pop(vcf.gl) <- pop_cinna

library(RColorBrewer)
library(pals)

myCol <- alphabet(n = 17) %>% setNames(popNames(vcf.gl))

msn <- poppr.msn(vcf.gl , distmat = bitwise.dist(vcf.gl ), palette = myCol, showplot = F)
set.seed(99)
plot_poppr_msn(vcf.gl , msn, inds = "na")


```

## DAPC

```{r, eval = F, fig.align="center", fig.width=12, fig.height=12}

library(poppr)

vcf.gl <- vcfR2genlight(vcf_cinna)
pop(vcf.gl) <- pop_cinna

mydapc <- dapc(vcf.gl, pop = pop(vcf.gl), parallel = F,  n.pca = 50, n.da = 5)

scatter.dapc(mydapc, clabel = 0.75, pch=15:20, scree.pca = TRUE, scree.da = FALSE, 
        posi.pca = "topright", posi.leg = "topleft", legend = F, 
        cleg = 0.9, inset.solid = 1, xax = 1, yax = 2, cex.lab = 1, cex = 2, solid = 1, cstar = 0)


```

## RAXML

Preparing input file for RAXML

```{r, eval = F, echo=T}

all.vcf <- vcf_cinna_sojae

gt.filtered <- extract.gt(all.vcf, element = "GT")
gt.filtered[gt.filtered == "0/0"] <- 0
gt.filtered[gt.filtered == "0/1"] <- 1
gt.filtered[gt.filtered == "1/0"] <- 1
gt.filtered[gt.filtered == "1/1"] <- 2
gt.filtered[gt.filtered == "./."] <- "?"
gt.filtered[is.na(gt.filtered)] <- "?"

gt.df <- apply(gt.filtered, 2, function (x) paste(x, collapse = ""))
gt.df <- gsub(pattern = "/", "", gt.df, fixed = T)

#write.table(gt.df, file = "205isolates.phy", quote = F)

```

## RAXML tree

```{r, eval=FALSE, fig.align="center", fig.height=12, fig.width=12, include=FALSE}

library(ape)

tree1 <- read.tree("./205Isolates/RAxML_bestTree.Ind205_results")
#tree1 <- read.tree("RAxML_bestTree.Pcinna_sojae_203")

id <- unlist(strsplit(tree1$tip.label, split = ".fq"))
pcinna_pop <- read.csv("New Microsoft Excel Worksheet.csv", header = TRUE)
pcinna_pop <- pcinna_pop[pcinna_pop$Isolate %in% id, ]
pcinna_pop <- pcinna_pop[match(id, pcinna_pop$Isolate), ]

pcinna_pop <- pcinna_pop[-nrow(pcinna_pop),]

pop <- as.character(pcinna_pop$Country)
pop[205] <- "Psojae"

pop <- as.factor(pop)

library(ggtree)
countryInfo <- split(tree1$tip.label, pop)
tree2 <- groupOTU(tree1, countryInfo)

tree2$tip.label <- as.character(pop)

ggtree(tree2, aes(color=group, label = node), layout="circular", ladderize = TRUE) + 
  geom_tiplab(size=3, aes(angle=angle))


```


```{r, fig.align="center", fig.height=20, fig.width=12}
ggtree(tree2, aes(color=group, label = node), ladderize = TRUE) +  geom_tiplab(size=3)
```


## Reconstructing ancestral states

```{r, eval = F}

library(ape)
mytree <- read.tree("./205Isolates/RAxML_bestTree.Ind205_results")
mytree <- drop.tip(mytree, "Psojae")

id <- unlist(strsplit(mytree$tip.label, split = ".fq"))
pcinna_pop <- read.csv("New Microsoft Excel Worksheet.csv", header = TRUE)
pcinna_pop <- pcinna_pop[pcinna_pop$Isolate %in% id, ]
pcinna_pop <- pcinna_pop[match(id, pcinna_pop$Isolate), ]
pop_cinna <- pcinna_pop$Country

tips <- unlist(strsplit(mytree$tip.label, split = ".fq"))
tips <- unlist(lapply(strsplit(tips, "-"), function(x) x[2]))

#tips <- paste(tips, pop_cinna, sep = "_")
#which(duplicated(tips))

len <- length(which(duplicated(tips)))
tips[which(duplicated(tips))] <- paste0(seq(1:len), tips[which(duplicated(tips))])

mytree$tip.label <- tips

ids <- as.data.frame(mytree$tip.label)

ids[2] <- pop_cinna
rownames(ids) <- ids$`mytree$tip.label`
ids <- ids[2]
ids <- as.matrix(ids)[, 1]

tree <- mytree
x <- ids

#plotTree(tree,type="fan",fsize=0.8,ftype="i", show.tip.label = F)
plot.phylo(ladderize(tree), type = "fan", show.tip.label = F)

library(pals)
library(phytools)
cols <- setNames(alphabet(n=17)[1:length(unique(x))],sort(unique(x)))

tiplabels(pie=to.matrix(x,sort(unique(x))),piecol=cols,cex=0.2)

#add.simmap.legend(colors=cols,prompt=F,x=0.9*par()$usr[1], y=-max(nodeHeights(tree)),fsize=0.8)

## Fitting ER model

fitER <- ace(x,tree,model = "ER", type="discrete")
fitER

#plotTree(ladderize(tree),type="fan",fsize=0.5,ftype="i")
plot.phylo(ladderize(tree), type = "fan", show.tip.label = F)

#plot.phylo(ladderize(tree),show.tip.label = F)

nodelabels(node=1:tree$Nnode+Ntip(tree),
           pie=fitER$lik.anc,piecol=cols,cex=0.3)

# tiplabels(pie=to.matrix(x,sort(unique(x))),piecol=cols,cex=0.1)

add.simmap.legend(colors=cols,prompt=TRUE,x=0.9*par()$usr[1],
                  y=-max(nodeHeights(tree)),fsize=0.9)


rspp <- as.data.frame(fitER$lik.anc[1, ])
colnames(rspp) <- "RSPP"
rspp <- cbind(pop = rownames(rspp), rspp)
rownames(rspp) <- NULL

library(ggplot2)

ggplot(data = rspp) + aes(x = pop, y = RSPP, fill = pop) + geom_bar(stat = "identity") 
  

```





## ADMIXTURE analysis

Preparing vcf file for admixture analysis

```{r, eval = F}

vcf
pop_cinna
in_pop <- c("Portugal", "Taiwan", "Australia", "South Africa", "USA", "Vietnam", "Chile", "Italy", "Dominican Republic")

library(mypackage)

#sublist_vcfbypop(vcf, pop_cinna)

vcf_9pop <- vcf_popsub(vcf = vcf, pop = pop_cinna, in_pop = in_pop)

#write.vcf(vcf_9pop, "vcf_9pop.gz")


```


### Plotting cross validation error

```{r Cross validaion error, echo=TRUE, fig.align="center", fig.width=12, fig.height=12}

library(ggplot2)
cv <- read.table(file.path("./MIN_5SAMPLES_9POP/CV_error.txt")) 
cv <- cv[,c(3:4)]
colnames(cv) <- c("K", "Cross_Validation_Error")
cv <- cv[order(readr::parse_number(cv$K)), ]
cv$K <- factor(cv$K, levels = cv$K)

ggplot(data = cv, aes(K, Cross_Validation_Error)) + geom_point(size  = 5) +
  theme(axis.text.x = element_text(size = 15)) +
  theme(axis.text.y = element_text(size = 15))

```


This is very similar to STRUCTURE plot.

```{r, admixture, echo=TRUE, fig.align="center", fig.width=18, fig.height=12}

library(reshape2)
qmat_files <- list.files("./MIN_5SAMPLES_9POP/qmatrices/")

source("admix_plot.R")

plot_list <- vector("list", length(qmat_files))

x <- read.vcfR("vcf_9pop.gz")

id <- unlist(strsplit(colnames(x@gt)[-1], split = ".fq"))
pcinna_pop <- read.csv("New Microsoft Excel Worksheet.csv", header = TRUE)
pcinna_pop <- pcinna_pop[pcinna_pop$Isolate %in% id, ]
pcinna_pop <- pcinna_pop[match(id, pcinna_pop$Isolate), ]

pop <- pcinna_pop$Country

for (k in 1:length(qmat_files)) {
    qmat <- read.table(file.path("./MIN_5SAMPLES_9POP/qmatrices/", qmat_files[k]))
    colnames(qmat) <- paste("Group", seq(1:ncol(qmat)), sep = ".")

    qmat <- cbind(pop, qmat)
    i <- sapply(qmat, is.factor)
    qmat[i] <- lapply(qmat[i], as.character)

    
    key <-  c("Australia" = "AUS", "Chile"= "CHL", "Dominican Republic" ="DOMREP",
              "Italy"= "ITL", "Portugal"= "PORT",
              "South Africa"= "SA","USA"= "USA",
              "Taiwan"= "TWN","Vietnam"= "VNM")
    
    qmat$pop <- factor(key[qmat$pop], ordered = TRUE, levels = key)
    
    qmat$pop <- factor(key[qmat$pop], levels = unique(qmat$pop)[c(2, 6, 3,  9, 1, 7, 5, 4, 8)], ordered = T)
     
    temp_plot <- admix_plot(qmat, horiz = F)
    temp_plot <- temp_plot + theme(axis.text.x = element_text(angle = 0, size = 15)) #+ theme_classic() 

    plot_list[[k]] <- temp_plot
    
} 

qmat_files    

library(cowplot)

  
plot_grid(plotlist = plot_list[14:16], nrow = 3, ncol = 1)  


```

## Index of association for Taiwan, Vietnam and Australian population


```{r, echo=FALSE, fig.align="center", fig.height=12, fig.width=18}

vcf.gl <- vcfR2genlight(vcf_cinna)
pop(vcf.gl) <- pop_cinna

t <- seppop(vcf.gl)

t1 <- clonecorrect(as.snpclone(t$Taiwan))

set.seed(100)
sex <- glSim(n.ind = nInd(t1), n.snp.nonstruc = ceiling(0.9*nLoc(t1)), n.snp.struc = floor(0.1*nLoc(t1)), ploidy=2, LD=TRUE)
### Structure (clonal pops)
clone <- glSim(nInd(t1), n.snp.nonstruc = floor(0.1*nLoc(t1)), n.snp.struc=ceiling(0.9*nLoc(t1)), ploidy=2, LD = T)
### Semi-clonal 
semi_clone <- glSim(nInd(t1),n.snp.nonstruc = 0.5*nLoc(t1), n.snp.struc= 0.5*nLoc(t1), ploidy=2, LD=T)
### Most-clonal 
most_clone <- glSim(nInd(t1), n.snp.nonstruc = ceiling(nLoc(t1)/3), n.snp.struc=2*nLoc(t1)/3, ploidy=2, LD=T)

## IA sex
ia.sex <- samp.ia(sex,quiet = T, reps = 100, n.snp = 100)
## IA clone
ia.clone <- samp.ia(clone, quiet = T, reps = 100, n.snp = 100)
## IA.semiclone
ia.semi <- samp.ia(semi_clone, quiet = T,reps = 100, n.snp = 100)
## IA.mostclone
ia.most <- samp.ia(most_clone, quiet = T, reps = 100, n.snp = 100)

ia.cinna <- samp.ia(t1,  reps = 100, quiet = T, n.snp = 100)

# Summarizing data frames
d1 <- data.frame(ia.cinna, rep("dataset", length(ia.cinna)))
d2 <- data.frame(ia.sex, rep("sexual", length(ia.sex)))
d3 <- data.frame(ia.clone, rep("clone", length(ia.clone)))
d4 <- data.frame(ia.semi, rep("semi-clone", length(ia.semi)))
d5 <- data.frame(ia.most, rep("most-clone", length(ia.semi)))
colnames(d1) <- c("ia","dset")
colnames(d2) <- c("ia","dset")
colnames(d3) <- c("ia","dset")
colnames(d4) <- c("ia","dset")
colnames(d5) <- c("ia","dset")
ia.total <- rbind(d3, d5, d4, d2, d1)
#ia.total <- rbind(d1, d2, d3, d4, d5)

# Normality tests
frames <- list(as.data.frame(d1), as.data.frame(d2), as.data.frame(d3), as.data.frame(d4), as.data.frame(d5))
normality <- list()
for (i in 1:length(frames)){
  normality[[i]] <- shapiro.test(frames[[i]][,'ia'])
}

# Analysis of variance
anova.ia <- aov(lm(ia ~ dset, ia.total))
library(agricolae)
tukey <- HSD.test(anova.ia, "dset", alpha = 0.001)
#tukey
# Kluskal wallis test
#kruskal.test(ia ~ dset, ia.total), trt="dset")
k.test <- with(ia.total, kruskal(ia, dset, group = T, p.adj = "bon"))

# Plot
 theme_set(theme_gray())
tai_IA <- ggplot(ia.total,aes(dset,ia,fill=dset)) + geom_boxplot() + xlab("Dataset") + ylab("Index of association") + ggtitle("Taiwan") + theme(plot.title = element_text(size = 16, face = "bold")) + theme(axis.text=element_text(size=14, face = "bold"), axis.title=element_text(size=16,face="bold"))
  



## Vietnam

vcf.gl <- vcfR2genlight(vcf_cinna)
pop(vcf.gl) <- pop_cinna

t <- seppop(vcf.gl)

t1 <- clonecorrect(as.snpclone(t$Vietnam))

set.seed(100)
sex <- glSim(n.ind = nInd(t1), n.snp.nonstruc = ceiling(0.9*nLoc(t1)), n.snp.struc = floor(0.1*nLoc(t1)), ploidy=2, LD=TRUE)
### Structure (clonal pops)
clone <- glSim(nInd(t1), n.snp.nonstruc = floor(0.1*nLoc(t1)), n.snp.struc=ceiling(0.9*nLoc(t1)), ploidy=2, LD = T)
### Semi-clonal 
semi_clone <- glSim(nInd(t1),n.snp.nonstruc = 0.5*nLoc(t1), n.snp.struc= 0.5*nLoc(t1), ploidy=2, LD=T)
### Most-clonal 
most_clone <- glSim(nInd(t1), n.snp.nonstruc = ceiling(nLoc(t1)/3), n.snp.struc=2*nLoc(t1)/3, ploidy=2, LD=T)

## IA sex
ia.sex <- samp.ia(sex,quiet = T, reps = 100, n.snp = 100)
## IA clone
ia.clone <- samp.ia(clone, quiet = T, reps = 100, n.snp = 100)
## IA.semiclone
ia.semi <- samp.ia(semi_clone, quiet = T,reps = 100, n.snp = 100)
## IA.mostclone
ia.most <- samp.ia(most_clone, quiet = T, reps = 100, n.snp = 100)

ia.cinna <- samp.ia(t1,  reps = 100, quiet = T, n.snp = 100)

# Summarizing data frames
d1 <- data.frame(ia.cinna, rep("dataset", length(ia.cinna)))
d2 <- data.frame(ia.sex, rep("sexual", length(ia.sex)))
d3 <- data.frame(ia.clone, rep("clone", length(ia.clone)))
d4 <- data.frame(ia.semi, rep("semi-clone", length(ia.semi)))
d5 <- data.frame(ia.most, rep("most-clone", length(ia.semi)))
colnames(d1) <- c("ia","dset")
colnames(d2) <- c("ia","dset")
colnames(d3) <- c("ia","dset")
colnames(d4) <- c("ia","dset")
colnames(d5) <- c("ia","dset")
ia.total <- rbind(d3, d5, d4, d2, d1)
#ia.total <- rbind(d1, d2, d3, d4, d5)

# Normality tests
frames <- list(as.data.frame(d1), as.data.frame(d2), as.data.frame(d3), as.data.frame(d4), as.data.frame(d5))
normality <- list()
for (i in 1:length(frames)){
  normality[[i]] <- shapiro.test(frames[[i]][,'ia'])
}

# Analysis of variance
anova.ia <- aov(lm(ia ~ dset, ia.total))
library(agricolae)
tukey <- HSD.test(anova.ia, "dset", alpha = 0.001)
#tukey
# Kluskal wallis test
#kruskal.test(ia ~ dset, ia.total), trt="dset")
k.test <- with(ia.total, kruskal(ia, dset, group = T, p.adj = "bon"))

# Plot
 theme_set(theme_gray())
vietnam_IA <- ggplot(ia.total,aes(dset,ia,fill=dset)) + geom_boxplot() + xlab("Dataset") + ylab("Index of association") + ggtitle("Vietnam") + theme(plot.title = element_text(size = 16, face = "bold")) + theme(axis.text=element_text(size=14, face = "bold"), axis.title=element_text(size=16,face="bold"))
  


## Australia


vcf.gl <- vcfR2genlight(vcf_cinna)
pop(vcf.gl) <- pop_cinna

t <- seppop(vcf.gl)

t1 <- clonecorrect(as.snpclone(t$Australia))

set.seed(100)
sex <- glSim(n.ind = nInd(t1), n.snp.nonstruc = ceiling(0.9*nLoc(t1)), n.snp.struc = floor(0.1*nLoc(t1)), ploidy=2, LD=TRUE)
### Structure (clonal pops)
clone <- glSim(nInd(t1), n.snp.nonstruc = floor(0.1*nLoc(t1)), n.snp.struc=ceiling(0.9*nLoc(t1)), ploidy=2, LD = T)
### Semi-clonal 
semi_clone <- glSim(nInd(t1),n.snp.nonstruc = 0.5*nLoc(t1), n.snp.struc= 0.5*nLoc(t1), ploidy=2, LD=T)
### Most-clonal 
most_clone <- glSim(nInd(t1), n.snp.nonstruc = ceiling(nLoc(t1)/3), n.snp.struc=2*nLoc(t1)/3, ploidy=2, LD=T)

## IA sex
ia.sex <- samp.ia(sex,quiet = T, reps = 100, n.snp = 100)
## IA clone
ia.clone <- samp.ia(clone, quiet = T, reps = 100, n.snp = 100)
## IA.semiclone
ia.semi <- samp.ia(semi_clone, quiet = T,reps = 100, n.snp = 100)
## IA.mostclone
ia.most <- samp.ia(most_clone, quiet = T, reps = 100, n.snp = 100)

ia.cinna <- samp.ia(t1,  reps = 100, quiet = T, n.snp = 100)

# Summarizing data frames
d1 <- data.frame(ia.cinna, rep("dataset", length(ia.cinna)))
d2 <- data.frame(ia.sex, rep("sexual", length(ia.sex)))
d3 <- data.frame(ia.clone, rep("clone", length(ia.clone)))
d4 <- data.frame(ia.semi, rep("semi-clone", length(ia.semi)))
d5 <- data.frame(ia.most, rep("most-clone", length(ia.semi)))
colnames(d1) <- c("ia","dset")
colnames(d2) <- c("ia","dset")
colnames(d3) <- c("ia","dset")
colnames(d4) <- c("ia","dset")
colnames(d5) <- c("ia","dset")
ia.total <- rbind(d3, d5, d4, d2, d1)
#ia.total <- rbind(d1, d2, d3, d4, d5)

# Normality tests
frames <- list(as.data.frame(d1), as.data.frame(d2), as.data.frame(d3), as.data.frame(d4), as.data.frame(d5))
normality <- list()
for (i in 1:length(frames)){
  normality[[i]] <- shapiro.test(frames[[i]][,'ia'])
}

# Analysis of variance
anova.ia <- aov(lm(ia ~ dset, ia.total))
library(agricolae)
tukey <- HSD.test(anova.ia, "dset", alpha = 0.001)
#tukey
# Kluskal wallis test
#kruskal.test(ia ~ dset, ia.total), trt="dset")
k.test <- with(ia.total, kruskal(ia, dset, group = T, p.adj = "bon"))

# Plot
 theme_set(theme_gray())
aus_IA <- ggplot(ia.total,aes(dset,ia,fill=dset)) + geom_boxplot() + xlab("Dataset") + ylab("Index of association")+
          ggtitle("Australia") + theme(plot.title = element_text(size = 16, face = "bold")) + theme(axis.text=element_text(size=14, face = "bold"), axis.title=element_text(size=16,face="bold"))
  

plot_grid(tai_IA, vietnam_IA, aus_IA, nrow = 2, ncol = 2)

```



## SNAPP  test

```{r}

source("R/VCF_SNAPP.R")


# library(vcfR)
# 
# vcf <- read.vcfR("Min10x_cov_203isolates_1027Variants.gz")
# 
# 
# id <- unlist(strsplit(colnames(vcf@gt)[-1], split = ".fq"))
# pcinna_pop <- read.csv("New Microsoft Excel Worksheet.csv", header = TRUE)
# pcinna_pop <- pcinna_pop[pcinna_pop$Isolate %in% id, ]
# pcinna_pop <- pcinna_pop[match(id, pcinna_pop$Isolate), ]
# pop_cinna <- pcinna_pop$Country
# 
# pop <- pop_cinna
# 
# 
# 
# in_pop <- c("Portugal", "Taiwan", "Australia", "USA", "Vietnam", "Chile", "South Africa")
# 
# 
# vcf_popsub(vcf, pop, in_pop)


source("R/vcf_popsub.R")

vcf <- vcf_popsub(vcf, pop, in_pop)
colnames(vcf@gt)

```

