---
title: "P.cinnamomi GBS analysis using TASSEL output"
author: "Shankar K Shakya"
date: "May 9, 2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.width = 18, fig.height = 12, echo = F, warning = F, message = F, verbose = F)

```

### Phytophtora cinnamomi GBS analysis


## Filtering samples that are not P.cinnamomi

The vcf file was generated using TASSEL pipeline by CGRB. Some of the samples in the provided vcf file were not P.cinnamomi so have to be removed. We have 144 P.cinnamomi isolates.

```{r, eval=FALSE, include=FALSE}
rm(list =ls())

library(vcfR, quietly = TRUE)
vcf <- read.vcfR("c1.mergedsnps.vcf-deconcatenate.vcf", verbose = FALSE)

gt <- vcf@gt
gt <- gt[, c(1:102, 104:146)]
vcf@gt <- gt

#head(vcf)
vcf


library(reshape2)
library(ggplot2) 
library(cowplot)

dp <- extract.gt(vcf, element = "DP", as.numeric=TRUE)


dpf <- melt(dp, varnames=c('Index', 'Sample'), value.name = 'Depth', na.rm=TRUE)
dpf <- dpf[ dpf$Depth > 0,]

samps_per_row <- 16
myRows <- ceiling(length(levels(dpf$Sample))/samps_per_row)
myList <- vector(mode = "list", length = myRows)

for(i in 1:myRows){
  myIndex <- c(i*samps_per_row - samps_per_row + 1):c(i*samps_per_row)
  myIndex <- myIndex[myIndex <= length(levels(dpf$Sample))]
  myLevels <- levels(dpf$Sample)[myIndex]
  myRegex <- paste(myLevels, collapse = "$|^")
  myRegex <- paste("^", myRegex, "$", sep = "")
  myList[[i]] <- dpf[grep(myRegex, dpf$Sample),]
  myList[[i]]$Sample <- factor(myList[[i]]$Sample)
}

# Create the plot.
myPlots <- vector(mode = "list", length = myRows)
for(i in 1:myRows){
  myPlots[[i]] <- ggplot(myList[[i]], aes(x=Sample, y=Depth)) + 
                  geom_violin(fill="#8dd3c7", adjust=1.0, scale = "count", trim=TRUE)

  myPlots[[i]] <- myPlots[[i]] + theme_bw()
  myPlots[[i]] <- myPlots[[i]] + theme(axis.title.x = element_blank(), 
                  axis.text.x = element_text(angle = 60, hjust = 1))
  myPlots[[i]] <- myPlots[[i]] + scale_y_continuous(trans=scales::log2_trans(), 
                  breaks=c(1, 10, 100, 800),
                  minor_breaks=c(1:10, 2:10*10, 2:8*100))
  myPlots[[i]] <- myPlots[[i]] + theme( panel.grid.major.y=element_line(color = "#A9A9A9", size=0.6) )
  myPlots[[i]] <- myPlots[[i]] + theme( panel.grid.minor.y=element_line(color = "#C0C0C0", size=0.2) )
}

plot_grid(plotlist = myPlots, nrow = myRows)



```


## Filtering samples by depth

```{r}

dp <- extract.gt(vcf, element = "DP", as.numeric=TRUE)
#dp <- dp[1:10, 1:10]
quants <- apply(dp, MARGIN=2, quantile, probs=c(0.1, 0.8), na.rm=TRUE)
dp2 <- sweep(dp, MARGIN=2, FUN = "-", quants[1,])
dp[dp2 < 0] <- NA

dp2 <- sweep(dp, MARGIN=2, FUN = "-", quants[2,])
dp[dp2 > 0] <- NA

dp[dp < 4] <- NA

vcf@gt[,-1][ is.na(dp) == TRUE ] <- NA


vcf ## 48.1% missing data

#heatmap.bp(dp[1:1000,], rlabels = FALSE)
```

## Omitting samples

```{r, eval=FALSE, include=FALSE}

myMiss <- apply(dp, MARGIN = 2, function(x){ sum( is.na(x) ) } )
myMiss <- myMiss / nrow(dp)
vcf@gt <- vcf@gt[, c(TRUE, myMiss < 0.6)]
vcf 

```

## Omitting variants


```{r, eval=FALSE, include=FALSE}

myMiss <- apply(dp, MARGIN = 1, function(x){ sum( is.na(x) ) } )
myMiss <- myMiss / ncol(dp)
vcf <- vcf[myMiss < 0.2, ]
vcf 



write.vcf(vcf, file = "Filtered.vcf.gz")
# dp <- extract.gt(vcf, element = "DP", as.numeric=TRUE)
# heatmap.bp(dp[1:1000,], rlabels = FALSE)

vcf <- read.vcfR("Filtered.vcf.gz")

```

## Filtering by Minor allele frequency

```{r, eval=FALSE, include=FALSE}

mymaf <- maf(vcf, element = 2)
head(mymaf)

nrow(mymaf)

mymaf <- mymaf[mymaf[,4] > 0.05,]


head(rownames(mymaf))
length(rownames(mymaf))

class(mymaf)

# variants <- strsplit(rownames(mymaf), "_")
# variants <- sapply(variants, function(x) x[2])
# head(variants)
# length(variants) #19057


nrow(vcf@fix) # 29521
head(vcf@fix)

fix <- vcf@fix[vcf@fix[,3] %in% rownames(mymaf), ]

nrow(fix)

vcf@fix <- fix


nrow(vcf@fix)

true_ind <- which(vcf@fix[,3] %in% rownames(mymaf))
vcf@fix <- vcf@fix[true_ind, ]

vcf@gt <- vcf@gt[true_ind, ]

write.vcf(vcf, file = "MAF_Filtered.vcf.gz")

```

## Depth Plot

```{r}

vcf <- read.vcfR("MAF_Filtered.vcf.gz")
dp <- extract.gt(vcf, element = "DP", as.numeric=TRUE)
par(mar=c(12,4,4,2))
boxplot(dp, col=2:8, las=3)
title(ylab = "Depth (DP)")

dp <- dp[,1:64]


```


## Converting VCFR object to genlight object


```{r}

library(adegenet)

vcf.gl <- vcfR2genlight(vcf)
# pca1 <- glPca(vcf.gl, parallel = FALSE)
# saveRDS(pca1, file = "mypca_MAF_Filtered_nf20.RData")


mypca <- readRDS("mypca_MAF_Filtered_nf20.RData")
pcasores <- mypca$scores

# rownames(pcasores)[grep("port", rownames(pcasores))] <- "Portugal"
# rownames(pcasores)[grep("ntaiw", rownames(pcasores))] <- "N.Taiwan"
# rownames(pcasores)[grep("staiw", rownames(pcasores))] <- "S.Taiwan"
# rownames(pcasores)[grep("centtaiw", rownames(pcasores))] <- "Cent.Taiwan"
# rownames(pcasores)[grep("nsw", rownames(pcasores))] <- "NSW"
# rownames(pcasores)[grep("safrica", rownames(pcasores))] <- "SAfrica"
# rownames(pcasores)[grep("nvietnam", rownames(pcasores))] <- "N.Vietnam"
# rownames(pcasores)[grep("waus", rownames(pcasores))] <- "W.Aus"
# rownames(pcasores)[grep("france", rownames(pcasores))] <- "France"
# rownames(pcasores)[grep("italy", rownames(pcasores))] <- "Italy"
# rownames(pcasores)[grep("domrep", rownames(pcasores))] <- "DomRep"
# rownames(pcasores)[grep("chile", rownames(pcasores))] <- "Chile"
# rownames(pcasores)[grep("newguin", rownames(pcasores))] <- "PNG"
# rownames(pcasores)[grep("tasaus", rownames(pcasores))] <- "Tasmania"
# rownames(pcasores)[grep("queensaus", rownames(pcasores))] <- "Queensland"
# rownames(pcasores)[grep("netaiw", rownames(pcasores))] <- "NE.Taiwan"

rownames(pcasores)[grep("port", rownames(pcasores))] <- "Portugal"
rownames(pcasores)[grep("taiw", rownames(pcasores))] <- "Taiwan"
rownames(pcasores)[grep("safrica", rownames(pcasores))] <- "SAfrica"
rownames(pcasores)[grep("france", rownames(pcasores))] <- "France"
rownames(pcasores)[grep("italy", rownames(pcasores))] <- "Italy"
rownames(pcasores)[grep("domrep", rownames(pcasores))] <- "DomRep"
rownames(pcasores)[grep("chile", rownames(pcasores))] <- "Chile"
rownames(pcasores)[grep("newguin", rownames(pcasores))] <- "PNG"
rownames(pcasores)[grep(c("queensaus|tasaus|waus|nswalesaus"), rownames(pcasores))] <- "Australia"
rownames(pcasores)[grep("nvietnam", rownames(pcasores))] <- "Vietnam"


df <- mypca$scores
df <- data.frame(Country = c(rownames(pcasores)), df)

library(ggfortify)

rownames(pcasores)[grep("Portugal|France|Italy", rownames(pcasores))] <- "Europe"
rownames(pcasores)[grep("Taiwan|Vietnam", rownames(pcasores))] <- "Asia"
rownames(pcasores)[grep("SAfrica", rownames(pcasores))] <- "Africa"
rownames(pcasores)[grep("Chile" , rownames(pcasores))] <- "South America"
rownames(pcasores)[grep("Australia" , rownames(pcasores))] <- "Australia"
rownames(pcasores)[grep("PNG" , rownames(pcasores))] <- "PNG"
rownames(pcasores)[grep("DomRep" , rownames(pcasores))] <- "DomRep"

df <- data.frame(Continent = c(rownames(pcasores)), df)

newdf <- df[,c(3,4)]

```

## PCA by country 

```{r}


pcaplot_country <- autoplot(prcomp(newdf), data = df, size = 4, colour = "Country")
pcaplot_country

```

## PCA by Continent

```{r}

pcaplot_contient <- autoplot(prcomp(newdf), data = df, size = 4, colour = "Continent")
pcaplot_contient

```

## DAPC plot

```{r}

pop(vcf.gl) <- df$Continent
dapc_continent <- dapc(vcf.gl, pop = pop(vcf.gl), n.pca=NULL, n.da=NULL, parallel = F, glPca = mypca)

saveRDS(dapc_continent, file = "MAF_mydapc_contient.RData", compress = FALSE)

dapc_continent <- readRDS("MAF_mydapc_contient.RData")

library(RColorBrewer)
myCol <- brewer.pal(nPop(vcf.gl), "Dark2") %>% setNames(popNames(vcf.gl))

scatter.dapc(dapc_continent, clabel = 0.75, pch=15:20, scree.pca = TRUE, scree.da = FALSE, 
        posi.pca = "topright", posi.leg = "topleft", legend = TRUE, 
        cleg = 0.9, inset.solid = 1, xax = 2, yax = 1, cex.lab = 1, cex = 2, solid = 1, cstar = 0)


pop(vcf.gl) <- df$Country
dapc_country <- dapc(vcf.gl, pop = pop(vcf.gl), n.pca=NULL, n.da=NULL, parallel = F, glPca = mypca)

saveRDS(dapc_country, file = "MAF_mydapc_country.RData", compress = FALSE)

dapc_country <- readRDS("MAF_mydapc_country.RData")

library(RColorBrewer)
myCol <- brewer.pal(nPop(vcf.gl), "Set3") %>% setNames(popNames(vcf.gl))

scatter.dapc(dapc_country,clabel = 0.75, pch=15:19, scree.pca = TRUE, scree.da = FALSE, 
        posi.pca = "topright", posi.leg = "topleft", legend = TRUE, 
        cleg = 0.9, inset.solid = 1, xax = 1, yax = 2, cex.lab = 1, cex = 1.5, solid = 1, cstar = 0)



pop(vcf.gl) <- df$Country
newpop <- popsub(vcf.gl, blacklist = c("Chile", "Vietnam", "Taiwan"))

newdapc <- dapc(newpop, pop = pop(newpop), n.pca = NULL, n.da = NULL, parallel = FALSE)


scatter.dapc(newdapc, clabel = 0.75, pch=15:19, scree.pca = TRUE, scree.da = FALSE, 
        posi.pca = "topright", posi.leg = "topleft", legend = TRUE, 
        cleg = 0.9, inset.solid = 1, xax = 2, yax = 1, cex.lab = 1, cex = 1.5, solid = 1, cstar = 0)



```


## Index of association

```{r IA, echo=TRUE}
library(poppr)

pop(vcf.gl) <- df$Country
ia_pop <- seppop(vcf.gl) %>% lapply(samp.ia) 

par(mfrow = c(2,5))
plot_list <- lapply(names(ia_pop), function(x) hist(ia_pop[[x]], main=x))


#mean <- lapply(ia_pop, function(x) mean(na.omit(x)))



```


## Minimum spanning network

```{r MSN, echo=TRUE}

library(RColorBrewer)

pop(vcf.gl) <- df$Continent
# myCol <- brewer.pal(nPop(vcf.gl), "Dark2") %>% setNames(popNames(vcf.gl))
msn <- poppr.msn(vcf.gl, distmat = bitwise.dist(vcf.gl), palette = rainbow)

set.seed(99)
plot_poppr_msn(vcf.gl, msn, inds = "nepal")


library(pals)
pop(vcf.gl) <- df$Country

myCol <- glasbey(n = 10) %>% setNames(popNames(vcf.gl))
msn <- poppr.msn(vcf.gl, distmat = bitwise.dist(vcf.gl), palette = myCol)
set.seed(99)
plot_poppr_msn(vcf.gl, msn, inds = "nepal")


```




## Allele frequency distribution

```{r, eval=FALSE, include=FALSE}

myFreq <- glMean(vcf.gl)
myFreq <- c(myFreq, 1-myFreq)
hist(myFreq, proba=TRUE, col="darkseagreen3", xlab="Allele frequencies",
main="Distribution of allele frequencies", nclass=20)
temp <- density(myFreq, bw=.05)
lines(temp$x, temp$y*2,lwd=3)


library(ape)


tre <- nj(dist(as.matrix(vcf.gl)))
tre

plot(tre, typ="fan", cex=0.7)



```



## Population differentiation

```{r, eval=FALSE, include=FALSE}

source("genetic_diff.R")

myDiff <- genetic_diff(vcf, df$Country)


# myDiff <- myDiff[myDiff$Hs_Australia > 0 & myDiff$Hs_Chile > 0 & myDiff$Hs_DomRep > 0 & myDiff$Hs_France > 0 & myDiff$Hs_Italy > 0 & myDiff$Hs_PNG > 0 & myDiff$Hs_Portugal > 0 & myDiff$Hs_SAfrica > 0 & myDiff$Hs_Taiwan > 0 & myDiff$Hs_Vietnam > 0 , ]
# 

dpf <- melt(myDiff[ , c("CHROM", "Gst", "Gprimest")] , na.rm=TRUE)

dpf <- melt(myDiff[ , c(1, 3:12, 27)] , na.rm=TRUE)

p <- ggplot(dpf, aes(x=variable, y=value)) + geom_violin(fill="#8dd3c7")
p <- p + xlab("Population differentiation statistics")
p <- p + ylab("Value")
p <- p + theme_bw()
p


vcf_list <- vector("list", length(unique(df$Country)))
names(vcf_list) <- unique(df$Country)

grep_list <-  list("port", "taiw", c("queensaus|tasaus|waus|nswalesaus"), "safrica", "nvietnam", "newguin", "chile", "italy", "france", "domrep")

grep_list <- list(c("port|france|italy"), c("taiw|nvietnam"), c("queensaus|tasaus|waus|nswalesaus"), "safrica", "newguin", "chile", "domrep")


for (i in (1:length(vcf_list))) {
  pop <- names(vcf_list[i])
  vcf_pop <- vcf
  gt <- vcf_pop@gt
  cols <- grep(grep_list[[i]], colnames(gt))
    
  vcf_pop@gt <- gt[,c(1,cols)]
  
  vcf_list[[i]] <- vcf_pop


}

saveRDS(vcf_list, file = "vcf_list.RData", compress = FALSE)


vcf_list <- readRDS("vcf_list.RData")

Pairs <- function(vcf_list) {
  pop <- names(vcf_list)
  pairs <- t(combn(pop,2))
  colnames(pairs) <- c("pop1", "pop2")
  as.data.frame(pairs, stringsAsFactors = FALSE)
}

pop.pairs <- Pairs(vcf_list)

pop.pairs$Pop_Combination <- apply(pop.pairs[, c("pop1", "pop2")], 1, paste, collapse = "-")

out <- vector("list", nrow(pop.pairs))
names(out) <- pop.pairs$Pop_Combination

temp.vcf <- vcf_list[[1]]

source("genetic_diff.R")
for (i in 1:nrow(pop.pairs)) {
  
  pair <- unlist(pop.pairs[i, ])
  
  pop1 <- pair[1]
  names(pop1) <- NULL
  vcf1 <- vcf_list[[grep(pop1, names(vcf_list))]]
  samp_num1 <- ncol(vcf1@gt) -1
  pop_len1 <- rep(pop1, samp_num1)
  
  
  pop2 <- pair[2]
  names(pop2) <- NULL
  vcf2 <- vcf_list[[grep(pop2, names(vcf_list))]]
  samp_num2 <- ncol(vcf2@gt) -1
  pop_len2 <- rep(pop2, samp_num2)
  
  pop <- c(pop_len1, pop_len2)
  
  
  temp.vcf@gt <- cbind(vcf1@gt, vcf2@gt[, -1])
  
  myDiff <- genetic_diff(temp.vcf, pops = factor(pop))
  out[[i]] <- myDiff
  
  
}


pairwise_diff_mat <- function(out) {

pairwise_gst_list <- lapply(out, function(x) colMeans(x[11], na.rm = TRUE))
pop.pairs$Gstprime <- unlist(pairwise_gst_list)
pop.pairs

df <- pop.pairs[, 3:4]

mat <- matrix(NA, nrow = length(vcf_list), ncol = length(vcf_list))
rownames(mat) <- names(vcf_list)
colnames(mat) <- names(vcf_list)
mat[lower.tri(mat)] <- df$Gstprime
mat[upper.tri(mat)] <- t(mat)[upper.tri(mat)]
mat[is.na(mat)] <- 0

return(mat)

}

pairwise_mat <- pairwise_diff_mat(out)


library(ape)
myNj <- ape::nj(as.dist(pairwise_mat))
myNj$edge.length[myNj$edge.length < 0] <- 0

plot(myNj, edge.width = 2)
add.scale.bar()

plot(hclust(dist(pairwise_mat)))


library(reshape2)
pairwise_mat[upper.tri(pairwise_mat)] <- NA
lower_tri <- melt(pairwise_mat, na.rm = TRUE)

library(ggplot2)
ggheatmap <- ggplot(lower_tri, aes(x = Var1, y = Var2, fill = value)) + geom_tile(color = "white") +
  scale_fill_gradient(low = "green", high = "red" , space = "Lab", name="Pairwise GSTprime") + theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 1, size = 12, hjust = 1)) + coord_fixed() +
  labs(x = "Population", y = "Population") +
  theme(axis.text.y = element_text(size = 12)) +
  theme(axis.title = element_text(size = 12)) +
  geom_text(aes(label = round(value, 2)))

ggheatmap
















```


## Number of heterozygous positions

```{r Heterozygosity, echo=TRUE}

gt <- extract.gt(vcf)
hets <- is.het(gt, na_is_false = TRUE)

sum_het <- colSums(hets)

sum_het.df <- as.data.frame(sum_het)
sum_het.df <- cbind(Isolate = rownames(sum_het.df), Country = df$Country, Heterozygosity = sum_het.df)
rownames(sum_het.df) <- NULL

ggplot(sum_het.df, aes(x = Country, y = sum_het, fill = Country)) + geom_boxplot(outlier.shape = NA) + geom_jitter() +
  ggtitle("Heterozygosity") +
  labs(x = "Country", y = "Number of heterozygous position") +
  theme(axis.text.x = element_text(size =12, angle = 60, hjust = 1)) +
  theme(axis.text.y = element_text(size =12)) + theme(legend.position = "none") +
  theme(plot.title = element_text(size = 30)) +
  theme(axis.title = element_text(size = 25))

```
























```{r structure, eval=FALSE, include=FALSE}

library(strataG)
Sys.setenv(PATH = paste("C:/Program Files (x86)/Structure2.3.4/bin", Sys.getenv("PATH"), sep = ";"))
Sys.setenv(PATH = paste("C:/Users/Shankar/Downloads/CLUMPP_Windows.1.1.2/", Sys.getenv("PATH"), sep = ";"))
vcf.gtype <- genind2gtypes(vcf_genind)


sr <- structureRun(vcf.gtype, k.range  = 1:4, burnin = 500, num.k.rep = 1, noadmix = F, numreps = 1000)

evno <- evanno(sr, plot = T)

qmat2 <- clumpp(sr, k = 2, sim.stat = "g.prime", repeats = 10)
plot(qmat2)

```

